### PR TITLE
Adding update and delete APIs for imageType image management API

### PR DESCRIPTION
**Pull request by utk-12**
Adding image management APIs for updating and deleting.

**_Query Params for POST Request:_**
session.id = "Insert session id from browser here"
updateImageOwners = "removeImageOwners OR addImageOwners"

**_Body for POST Request (Required Parameters):_**

```
{
"imageType": "utktest",
"ownerships":[{"owner": "jakhani", "role": "ADMIN"},
{"owner": "djaiswal", "role": "ADMIN"}]
}
```

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java [hunk 1]
```diff
@@ -106,22 +124,82 @@ public int createImageType(final ImageType imageType) {
...
   }
 
+  @Override
+  public int removeOwnerOfImageType(ImageType imageType) throws ImageMgmtException {
+    final Set<ImageOwnership> currentOwners =
+        new HashSet<>(getImageTypeOwnership(imageType.getName()));
+    final Set<ImageOwnership> ownersToRemove = new HashSet<>(imageType.getOwnerships());
+    ownersToRemove.retainAll(currentOwners);
+      //Check if it's removing all owners
+    if (ownersToRemove.size() >= currentOwners.size() - 1){
```
Comment (amielkelinkedin) [inline azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:177]: Is this actually checking if there are two members? It seems like it is doing a size comparison.

File: azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java [hunk 2]
```diff
@@ -175,29 +175,41 @@ protected void handlePost(
...
+      genericImageType.setCreatedBy(session.getUser().getUserId());
+      if (!HttpRequestUtils.hasParam(req, "updateImageOwners")){
+        if (!hasImageManagementPermission(imageType, session.getUser(), Type.CREATE)) {
+          log.debug(String.format("Invalid permission to create image type for "
+              + "user: %s, image type: %s.", session.getUser().getUserId(), imageType));
+          throw new ImageMgmtInvalidPermissionException(ErrorCode.FORBIDDEN, "Invalid permission to "
+              + "create image type");
+        }
+        // Create image type /Update  and get image type id
+        imageTypeId = this.imageTypeService.createImageType(genericImageType);
```
Comment (amielkelinkedin) [inline azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java:189]: Can we change the name of this method from create to update?

File: azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java [hunk 3]
```diff
@@ -175,29 +175,41 @@ protected void handlePost(
...
             + " null. Must provide valid imageType to create/register image type.");
       }
-      if (!hasImageManagementPermission(imageType, session.getUser(), Type.CREATE)) {
-        log.debug(String.format("Invalid permission to create image type for "
-            + "user: %s, image type: %s.", session.getUser().getUserId(), imageType));
-        throw new ImageMgmtInvalidPermissionException(ErrorCode.FORBIDDEN, "Invalid permission to "
-            + "create image type");
+      final Integer imageTypeId;
+      genericImageType.setModifiedBy(session.getUser().getUserId());
+      genericImageType.setCreatedBy(session.getUser().getUserId());
```
Comment (amielkelinkedin) [inline azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java:180]: Wouldn't this overwrite the creator in the DB with the updater?

**Comment by amielkelinkedin** [comment]
Thank you for making this change! This is a really interesting approach!

Do we know how we want to structure this change? Here we are assuming a user will want to add and remove users as separate operations rather than just POSTing over an existing ImageOwnership record with a new request. I see pluses and minuses to both approaches. Here if the request is largely the same the user doesn't have to include a lot of the same information, which is quite nice. One the other hand, by simply POSTing over the object we could add and remove at the same time. It would also make updates backwards compatible to our current Jenkins job. If we're planning on moving towards LDAP, we might have smaller requests in the future with fewer users.

I'm fine with either approach!

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java [hunk 4]
```diff
@@ -106,22 +124,82 @@ public int createImageType(final ImageType imageType) {
...
+            imageType.getModifiedBy(), currentTimestamp);
       }
-      if (e.getErrorCode() == SQL_ERROR_CODE_DATA_TOO_LONG) {
-        errorMessage = "Reason: Data too long for one or more column(s).";
+      transOperator.getConnection().commit();
+      return 1;
+    };
+    try {
+      this.databaseOperator.transaction(update);
+      if (imageTypeId < 1) {
```
Comment (djaiswal83) [inline azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:157]: Why imagetypeId check matters here? the value is set at line 153. The update does not affect its value.

Comment (djaiswal83) [reply azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:157]: The imagetype check for validity must happen before the update SQL is executed.

Comment (djaiswal83) [reply azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:157]: The try catch block needs rearrangement by moving the transaction call after the if block and adding its own return value check.

**Comment by hqin2016** [comment]
> Thank you for making this change! This is a really interesting approach!
> 
> Do we know how we want to structure this change? Here we are assuming a user will want to add and remove users as separate operations rather than just POSTing over an existing ImageOwnership record with a new request. I see pluses and minuses to both approaches. Here if the request is largely the same the user doesn't have to include a lot of the same information, which is quite nice. One the other hand, by simply POSTing over the object we could add and remove at the same time. It would also make updates backwards compatible to our current Jenkins job. If we're planning on moving towards LDAP, we might have smaller requests in the future with fewer users.
> 
> I'm fine with either approach!

@utk-12 could you sync up with @amielkelinkedin offline regarding his suggestion?

**Review by hqin2016 — COMMENTED** [review]
Thanks for taking this story. This is indeed a very useful function.

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDao.java [hunk 5]
```diff
@@ -86,4 +86,21 @@
...
    */
   public ImageType getImageTypeWithOwnershipsById(String id);
+
+  /**
+   * Add an owner for a given image type.
+   *
+   * @param imageType - ImageType to be updated
+   * @return int - Id representing image type metadata
+   */
+  public int addOwnerOfImageType(ImageType imageType) throws ImageMgmtException;
```
Comment (hqin2016) [inline azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDao.java:96]: how about using addImageTypeOwner as the function name? the same for the method below.

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java [hunk 6]
```diff
@@ -106,22 +124,82 @@ public int createImageType(final ImageType imageType) {
...
+  }
+
+  @Override
+  public int addOwnerOfImageType(final ImageType imageType) {
+    final Set<ImageOwnership> currentOwners =
+        new HashSet<>(getImageTypeOwnership(imageType.getName()));
+    final Set<ImageOwnership> newOwners = new HashSet<>(imageType.getOwnerships());
+    newOwners.removeAll(currentOwners);
+    //Check if there are actually new owners added.
+    if (newOwners.isEmpty()) {
```
Comment (hqin2016) [inline azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:139]: can we add a log here?

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java [hunk 7]
```diff
@@ -106,22 +124,82 @@ public int createImageType(final ImageType imageType) {
...
+      //Check if it's removing all owners
+    if (ownersToRemove.size() >= currentOwners.size() - 1){
+        throw new ImageMgmtDaoException(ErrorCode.BAD_REQUEST, "Exception while deleting "
+            + " image owners, need greater than two owners to be present");
+      }
+      int imageTypeId =
+          getImageTypeByName(imageType.getName()).map(BaseModel::getId).orElse(0);
+
+    try {
+      if (imageTypeId < 1) {
```
Comment (hqin2016) [inline azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:185]: @djaiswal83 's comment above applies here too.

File: azkaban-common/src/main/java/azkaban/imagemgmt/dto/ImageTypeDTO.java [hunk 8]
```diff
@@ -65,6 +65,7 @@ public void setDeployable(final Deployable deployable) {
   }
 
   public void setOwnerships(final List<ImageOwnershipDTO> ownerships) {
-    this.ownerships = ownerships;
+            ownerships.stream().forEach(o -> o.setName(getName()));
```
Comment (hqin2016) [inline azkaban-common/src/main/java/azkaban/imagemgmt/dto/ImageTypeDTO.java:68]: Why is this needed?

File: azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java (hunk 3 above)
Comment (hqin2016) [inline azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java:180]: I thought createdBy and modifiedBy are different people. You might only need to set modifiedBy.

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java (hunk 1 above)
Comment (utk-12) [reply azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:177]: I can do that. I shall change it to get the final owner list size and see if it's greater than 1.

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDao.java (hunk 5 above)
Comment (utk-12) [reply azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDao.java:96]: ack

File: azkaban-common/src/main/java/azkaban/imagemgmt/dto/ImageTypeDTO.java (hunk 8 above)
Comment (utk-12) [reply azkaban-common/src/main/java/azkaban/imagemgmt/dto/ImageTypeDTO.java:68]: This is needed becuase the imageOwnershipDTO doesn't get its name field populated. It shows as "null" in the request.

File: azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java (hunk 3 above)
Comment (utk-12) [reply azkaban-web-server/src/main/java/azkaban/imagemgmt/servlets/ImageTypeServlet.java:180]: in the case of creating the image_types table maybe, but here either the person a create is happening where both will be the same. or a new entry in the image_ownership table takes place for which both created and modified entry will be the user logged in.

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java (hunk 4 above)
Comment (utk-12) [reply azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:157]: ack. Will change this.

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java (hunk 7 above)
Comment (utk-12) [reply azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:185]: It is re-arranged here right ? I presume this is the correct way ? where I check if the imageType id before the sql update.

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java [hunk 9]
```diff
@@ -106,22 +124,85 @@ public int createImageType(final ImageType imageType) {
...
 
+  @Override
+  public int removeImageTypeOwner(ImageType imageType) throws ImageMgmtException {
+    final Set<ImageOwnership> currentOwners =
+        new HashSet<>(getImageTypeOwnership(imageType.getName()));
+    final Set<ImageOwnership> ownersToRemove = new HashSet<>(imageType.getOwnerships());
+    ownersToRemove.retainAll(currentOwners);
+    currentOwners.removeAll(ownersToRemove);
+      //Check if it's removing all owners
+    if (!(currentOwners.size() > 1)){
```
Comment (hqin2016) [inline azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java:179]: nit: isEmpty() may be better

**Review by hqin2016 — APPROVED** [review]

**Review by djaiswal83 — CHANGES_REQUESTED** [review]
Please address the comments

File: azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java [hunk 10]
```diff
@@ -152,13 +153,13 @@ public int addOwnerOfImageType(final ImageType imageType) {
       transOperator.getConnection().commit();
       return 1;
     };
+    if (imageTypeId < 1) {
```
[the remark this answers is not in the document]
Comment (utk-12) [reply azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageTypeDaoImpl.java]: Addressed

**Review by djaiswal83 — APPROVED** [review]
Thanks for addressing the comments.
